Worker debug

eugeneo opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

This is a proposed interface for debugging Node workers. It is largely similar to Chrome DevTools protocol with web-specific parts removed.

eugeneo commented:
This pull request includes commit from https://github.com/nodejs/node/pull/21182

ak239 commented on src/inspector/node_protocol_config.json:
```diff
@@ -3,12 +3,7 @@
-        "namespace": ["node", "inspector", "protocol"],
-        "options": [
-            {
-                "domain": "NodeTracing"
```
Why this one is removed?

eugeneo commented:
> I am wondering: should we instrument worker postMessage and onmessage with V8Inspector instrumentation for step-into worker and async stacks cross worker boundary?

Out of scope for this PR, IMHO.

pavelfeldman commented on src/inspector/node_protocol.pdl:
```diff
@@ -37,3 +37,59 @@ experimental domain NodeTracing
+      string url
+
+  # Sends protocol message over session with given id.
+  command sendMessageToTarget
```
This does not exist in the new flattened world, sessionId attribute is passed as a top-level parameter instead. This allows us to not JSON-serialize and wrap all the commands that go into the nested targets such as workers.

eugeneo replied:
I can still see it here - https://paulirish.github.io/debugger-protocol-viewer/1-3/Target/

pavelfeldman replied:
We have not migrated all the use cases to flattened yet, it also is a public API method, so deprecation would beed to be graceful. Basically, it'll soon become deprecated and a year later it'll become noop.

eugeneo commented:
@dgozman had approved this protocol file when implementation had started.

pavelfeldman commented:
Well, since then we've introduced the flattened operation... Furthermore, we were considering moving the workers off the Target domain discovery, but that is still up in the air. Anyhow, let's see what he says.

pavelfeldman commented on src/inspector/node_protocol.pdl:
```diff
@@ -37,3 +37,59 @@ experimental domain NodeTracing
+      boolean autoAttach
+      # Whether to pause new targets when attaching to them. Use `Runtime.runIfWaitingForDebugger`
+      # to run paused targets.
+      boolean waitForDebuggerOnStart
```
As I see it, we have two ways of proceeding with the worker inspection:

1) Consider NodeTarget to be a duck-typed Target domain

This implies that flattened operation is in place. Flatten should not be hard for you to add, simply copy protocol from here: https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/inspector/browser_protocol.pdl?q=browser_protocol&sq=package:chromium&g=0&l=5950 and add corresponding sniffing logic. When we move the front-end to be flatten by default, it will 'just work'.

2) Name things with what they are and rename this domain to NodeWorker.

Basically rename all the 'target' mentions here with the corresponding 'worker' mentions. Then you don't need to add flattened operation, Target domain is special and it adds the sessionId aspect into the protocol, other domains, including NodeWorker would not be special, so they would still need to wrap the messages. That's fine. We would still be lying to ourselves because we would be creating a target on the front-end for each worker, but I think that is tolerable.

@dgozman: wdyt?

dgozman commented on src/inspector/node_protocol.pdl:
```diff
@@ -37,3 +37,59 @@ experimental domain NodeTracing
   event tracingComplete
+
+# Supports additional targets discovery and allows to attach to them.
+experimental domain NodeTarget
```
After some discussion, we think that we'll not be able to keep NodeTarget domain close to Chromium's Target domain on the frontend side in the long term. Therefore, it would be easier to keep this domain very node-worker-specific to reduce the complexity and maintenance for the Node community.

My proposal would be to just rename things, keeping semantics the same:
- NodeTarget -> NodeWorker
- TargetID -> WorkerID
- TargetInfo -> WorkerInfo
- sendMessageToTarget -> sendMessageToWorker
- setAutoAttach(true, waitForDebuggerOnStart) -> enable(waitForDebuggerOnStart)
- setAutoAttach(false) -> disable()
- attachedToTarget -> workerCreated
- detachedFromTarget -> workerDestroyed
- receivedMessageFromTarget -> receivedMessageFromWorker

What do you think? Other than that, things look good to me.

eugeneo replied:
Sounds like a plan! I will implement the changes.

eugeneo replied:
Did that. One thing I kinda feel strongly is that worker(Created|Destroyed) would be a misnomer, so I renamed it to attachedToWorker/detachedFromWorker. I would also consider naming them simply "attached" or "detached" as they are already have NodeWorker prefix.

dgozman replied:
Sounds good, I like "attachedToWorker/detachedFromWorker".

dgozman commented on src/inspector/node_protocol.pdl:
```diff
@@ -37,3 +37,58 @@ experimental domain NodeTracing
   # delivered via dataCollected events.
   event tracingComplete
+
+# Supports additional targets discovery and allows to attach to them.
```
This description should probably talk about workers as well.

eugeneo replied:
Done.

dgozman commented on src/inspector/node_protocol.pdl:
```diff
@@ -37,3 +37,58 @@ experimental domain NodeTracing
+  command enable
+    parameters
+      # Whether to pause new targets when attaching to them. Use `Runtime.runIfWaitingForDebugger`
+      # to run paused targets.
```
targets -> workers?

eugeneo replied:
Done.

eugeneo commented:
@dgozman @pavelfeldman qq: should worker inspectors have the NodeTarget domains?

dgozman commented:
@eugeneo If workers can spawn nested workers, it would be ideal to report them through the main NodeWorker domain if possible.

addaleax commented on src/inspector/worker_agent.cc:
```diff
@@ -0,0 +1,155 @@
+  void WorkerCreated(const std::string& title,
+                 const std::string& url,
+                 bool waiting,
+                 std::shared_ptr<MainThreadHandle> target);
```
Style nit: Here and above, the parameters aren’t aligned (and we generally only indent the initializers starting with `:` by 2 spaces.)

Side note: Running `make format-cpp` (preceded by `make format-cpp-build`, if necessary) would make sure you don’t have to worry about style nits anymore.

addaleax commented on src/inspector/worker_agent.cc:
```diff
@@ -0,0 +1,155 @@
+      .setWorkerId(id)
+      .setTitle(title)
+      // Placeholder URL still needed for some frontends to resolve paths
+      .setUrl(url.empty() ? "file://eval" : url)
```
Is it intentional that this has 2 slashes, i.e. the *host* is being set to `eval`?

eugeneo replied:
That's what Chrome DevTools team suggested, but they do not have equivalent feature in Chrome. It might be ok to set it to another value.

pavelfeldman replied:
> That's what Chrome DevTools team suggested

Hm. I would simply pass url as is. Empty is empty, that should be fine.

eugeneo replied:
Empty URL it is.

addaleax commented on src/inspector/worker_inspector.h:
```diff
@@ -0,0 +1,97 @@
+  bool WaitForConnect() {
+    return wait_;
+  }
+ private:
```
Style nit: blank line before `private:` (I think the linter should complain about this?)

addaleax commented on src/inspector_agent.cc:
```diff
@@ -606,11 +623,18 @@ class NodeInspectorClient : public V8InspectorClient {
+  if (start_io_thread_async.data == this) {
+    start_io_thread_async.data = nullptr;
+    // This is global, will never get freed
+    uv_close(reinterpret_cast<uv_handle_t*>(&start_io_thread_async), nullptr);
```
This is interesting – why is `start_io_thread_async` global? That doesn’t seem quite right (or thread-safe, for that matter). These things should probably per-Environment or per-Isolate?

eugeneo replied:
It is used by a signal handler.

eugeneo commented:
@addaleax Thank you for the review. I addressed the comments and rebased the patch to make sure it works on top of current head.

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16839/

eugeneo commented:
Looks like the CI is failing due to infrastructure...

eugeneo commented:
@addaleax I would really appreciate you taking a look at this code as it touches the workers implementation.

eugeneo commented:
@addaleax Should be good to merge. I rebased/squashed, fixed VC compilation error and updated the test to use the new URL API.

CI: https://ci.nodejs.org/job/node-test-pull-request/17004/

eugeneo commented:
Another CI attempt: https://ci.nodejs.org/job/node-test-pull-request/17007/

Older versions had one file with .cpp extension while the node.gyp used .cc. Interestingly enough, worked like a charm on *nixes but not the Windows...

eugeneo commented:
A number of unexpected CI failures, will need to investigate.

eugeneo commented:
Tests are failing because the code is being ran in a worker. I will do a triage later this week to see if there's anything actually broken and prevent tests that should not be running in a worker.

eugeneo commented:
Parts of this PR landed in a separate PR. I rebased this code. New CI: https://ci.nodejs.org/job/node-test-pull-request/17253/

eugeneo commented:
Landed as https://github.com/nodejs/node/commit/f28c6f7eef58e7c3133bb2cd457d05b986194ba3

targos commented:
Should this be backported to `v10.x-staging`? The change lands cleanly but the test fails:

```
$ out/Release/node --experimental-worker /home/mzasso/git/nodejs/v10.x/test/parallel/test-worker-debug.js
Test basic debug scenario
/home/mzasso/git/nodejs/v10.x/test/common/index.js:662
const crashOnUnhandledRejection = (err) => { throw err; };
                                             ^

AssertionError [ERR_ASSERTION]: Input A expected to strictly equal input B:
+ expected - actual

- 0
+ 2
    at WorkerSession.waitForBreakAfterCommand (/home/mzasso/git/nodejs/v10.x/test/parallel/test-worker-debug.js:111:12)
```

eugeneo commented:
I definitely see a lot of value in introducing this in the same Node version the workers are introduced. This difference should be due to some V8 change - I adjusted the test case and the test works - https://github.com/eugeneo/node/tree/v10-workers-debug

eugeneo commented:
@targos - sorry, forgot to tag you. I also did a CI run - seems ok - https://ci.nodejs.org/job/node-test-commit/21627/

targos commented:
@eugeneo Excellent! Would you like to open a backport PR?
It looks like you could just create one from that branch and target `v10.x-staging`. Feel free to ping me on it and link the CI run.
